src/src/com/microsoft/aad/adal/AuthenticationContext.java

```diff
@@ -1580,138 +1625,90 @@ private String getTokenHash(String token) {
-                Logger.v(TAG, "Looking for Multi Resource Refresh token");
-                keyUsed = CacheKey.createMultiResourceRefreshTokenKey(request, userId);
-                item = mTokenCacheStore.getItem(keyUsed);
-            }
+            Logger.v(TAG + methodName, "Look for resource specific refresh token in the cache.");
+            refreshItem = getRefreshTokenFromCache(KeyEntryType.REGULAR_REFRESH_TOKEN_ENTRY, request);
+            
+            // If regular RT is found, but it's not MRRT, shouldn't bother trying with FRT. All the token issued
+            // by EvoSts is MRRT. If regular RT is not MRRT, it basically means it not issued by EvoSts. FoCI feature
+            // is only visible to RT issued by evoSts. 
+            if (refreshItem == null || refreshItem.mMultiResource) {
+                final String regularRTLookupMsg = refreshItem == null ? "Resource specific RT does not exist, look for MRRT." 
+                        : "Found resource specific RT, which is also MRRT. Look for MRRT.";
+                Logger.v(TAG + methodName, regularRTLookupMsg);
+
+                final RefreshItem mrrtItem = getRefreshTokenFromCache(KeyEntryType.MULTI_RESOURCE_REFRESH_TOKEN_ENTRY, request);
+                refreshItem = mrrtItem;
+                
+                // If RT is found, but it's not marked as MRRT, shouldn't bother trying to find FRT. 
+                if (refreshItem == null || !StringExtensions.IsNullOrBlank(refreshItem.mFamilyClientId)) {
+                    final String mrrtLookupMsg = refreshItem == null ? "MRRT does not exist, look for FRT."
+                            : "Found MRRT which is also a FRT, look for FRT.";
 
-            // Temporarily disable the family client id feature. 
-            final boolean isFamilyClientIdFeatureEnabled = false;
-            if (isFamilyClientIdFeatureEnabled) {
-                // If still cannot find an item, try with family clientid
-                // Disable this feature for ADFS authority
-                if (item == null || StringExtensions.IsNullOrBlank(item.getRefreshToken())) {
-                    if (!UrlExtensions.isADFSAuthority(StringExtensions.getUrl(mAuthority))) {
-                        Logger.v(TAG + methodName, "No refresh token found, trying to find family client id item.");
-                        item = findFamilyItemForUser(request.getUserIdentifierType(), userId);
-                    } else {
-                        Logger.v(TAG + methodName, "No refresh token found, skip the family client id item lookup for ADFS authority.");
+                    Logger.v(TAG + methodName, mrrtLookupMsg);
+                    refreshItem = getRefreshTokenFromCache(KeyEntryType.FAMILY_REFRESH_TOKEN_ENTRY, request);
+                    
+                    if (refreshItem == null) {
+                        Logger.v(TAG + methodName, "FRT does not exist, fall back to MRRT if exists.");
+                        refreshItem = mrrtItem;
```

Comment:
Not quite sure I follow. The fallback method fallBackToMRRT is only called if we find a FRT and try with the FRT, and it fails.